src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -573,17 +576,33 @@ internal static bool Dispatch()
                         {
                             ThreadPool.ReportThreadStatus(isWorking: true);
                             reportedStatus = true;
-                            workItem.ExecuteWorkItem();
+                            if (workItem is IThreadPoolWorkItem wi)
+                            {
+                                wi.Execute();
+                            }
+                            else
+                            {
+                                Debug.Assert(workItem is Task);
+                                Unsafe.As<Task>(workItem).ExecuteFromThreadPool();
+                            }
                         }
                         finally
                         {
                             if (reportedStatus)
                                 ThreadPool.ReportThreadStatus(isWorking: false);
                         }
                     }
+                    else if (workItem is IThreadPoolWorkItem wi)
```

Comment:
Likewise here is Task first faster? (as is a concrete type)

Though can you Unsafe to an interface?

Reply:
Seems you can Unsafe.As to an interface; the asm doesn't look much different, just swapping the two blocks of the if; so it just depends on the difference in cast speed for the `isinst IThreadPoolWorkItem` vs `isinst Task`. 

Might not be meaningful...

Reply:
Thanks, I'll try, and will probably switch.  I preferred to keep the interface first under the assumption that the majority of work items going through are implementations of the interface rather than tasks, but thinking about it more, that may not actually be true... I expect it'll be true for delegate-backed tasks like Task.Run, but now that we queue the async method state machine box directly, and now that (as of this change) it's not implementing the interface, we're likely to actually see this take the task branch much more frequently, since every async state machine continuation (in most async methods where we optimize with that approach) is derived from task.

Reply:
Is there a case where you would derive from Task AND implement the interface? If so, which 'Execute' would you want to call?

Reply:
Not internally; I made sure of that. Technically externally you could, in which case we should prefer Task. It's already super rare to derive from task externally (only works in the delegate-backed case, in coordination with Start). And I don't think it's worth adding extra checks for all to protect against that case.

Reply:
I've swapped the order, as it is faster (in microbenchmarks) to check for Task and then Unsafe.As for the interface rather than vice versa, in particular when the object implements a bunch of other interfaces first, as that impacts the speed of the cast/type check but not of the Unsafe.As (which is a nop).  However, I have added a check for Task in the public API to throw if TP.UnsafeQueueUserWorkItem is passed a Task-derived type that implements the interface; all of the existing internal usage skips this.